### PR TITLE
[feature]FE: drop worker when drop a backend 

### PR DESCRIPTION
**Pull request by abc982627271**
## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

When drop a backend, fe need to removeWorker to starMgr, dropping the starlet on the backend.

### CODE REVIEW

**Comment by abc982627271** [comment]
run starrocks_be_unittest

**Comment by abc982627271** [comment]
run starrocks_be_unittest

**Comment by abc982627271** [comment]
run starrocks_be_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 11 / 19 (57.89%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/system/SystemInfoService.java|1|4|25.00%|
|:large_blue_circle:|com/starrocks/catalog/StarOSAgent.java|10|15|66.67%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 19 / 22 (86.36%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/StarOSAgent.java|14|17|82.35%|
|:large_blue_circle:|com/starrocks/system/SystemInfoService.java|5|5|100.00%|

File: fe/fe-core/src/main/java/com/starrocks/catalog/StarOSAgent.java [hunk 1]
```diff
@@ -148,6 +148,25 @@ public void addWorker(long backendId, String workerIpPort) {
         LOG.info("add worker {} success, backendId is {}", workerId, backendId);
     }
 
+    public void removeWorker(String workerIpPort) {
+        if (!workerToId.containsKey(workerIpPort)) {
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/catalog/StarOSAgent.java:152]: If FE has just started, maybe map is empty.
Should check if it exists in StarMgr, or directly drop from StarMgr.

File: fe/fe-core/src/main/java/com/starrocks/system/SystemInfoService.java [hunk 2]
```diff
@@ -263,6 +264,11 @@ public void dropBackend(String host, int heartbeatPort, boolean needCheckUnforce
         final Cluster cluster = GlobalStateMgr.getCurrentState().getCluster(droppedBackend.getOwnerClusterName());
         if (null != cluster) {
             cluster.removeBackend(droppedBackend.getId());
+            // remove worker
+            if (Config.integrate_starmgr) {
+                String starletHost = droppedBackend.getHost() + ":" + droppedBackend.getStarletPort();
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/system/SystemInfoService.java:269]: Starlet port may be not reported when FE has just started.

Comment (wyb) [reply fe/fe-core/src/main/java/com/starrocks/system/SystemInfoService.java:269]: What to do if remove worker fails?
Maybe we can try to remove worker first, and return error if remove worker fails.

File: fe/fe-core/src/main/java/com/starrocks/catalog/StarOSAgent.java (hunk 1 above)
Comment (wuleistarrocks) [reply fe/fe-core/src/main/java/com/starrocks/catalog/StarOSAgent.java:152]: agree with @wyb.  I think drop directly is ok :)

File: fe/fe-core/src/main/java/com/starrocks/catalog/StarOSAgent.java [hunk 3]
```diff
@@ -148,6 +148,30 @@ public void addWorker(long backendId, String workerIpPort) {
         LOG.info("add worker {} success, backendId is {}", workerId, backendId);
     }
 
+    public void removeWorker(String workerIpPort) throws StarClientException {
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/catalog/StarOSAgent.java:151]:
```suggestion
    public void removeWorker(String workerIpPort) throws DdlException {
```

Comment (wyb) [reply fe/fe-core/src/main/java/com/starrocks/catalog/StarOSAgent.java:151]: Better to not expose `StarClientException` outside.

File: fe/fe-core/src/main/java/com/starrocks/catalog/StarOSAgent.java [hunk 4]
```diff
@@ -148,6 +148,30 @@ public void addWorker(long backendId, String workerIpPort) {
...
+        } catch (StarClientException e2) {
+            LOG.warn(e2);
+            return;
+        }
+
+        try {
+            client.removeWorker(serviceId, workerId);
+        } catch (StarClientException e) {
+            if (e.getCode() != StarClientException.ExceptionCode.NOT_EXIST) {
+                throw new StarClientException(e.getCode(), "remove worker error");
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/catalog/StarOSAgent.java:166]: throw new DdlException("Failed to remove worker. error: " + e.getMessage());

File: fe/fe-core/src/main/java/com/starrocks/catalog/StarOSAgent.java [hunk 5]
```diff
@@ -148,6 +148,30 @@ public void addWorker(long backendId, String workerIpPort) {
...
+            }
+        }
+
+        workerToBackend.remove(workerId);
+        workerToId.remove(workerIpPort);
+        LOG.info("remove worker {} success in StarOSAgent", workerIpPort);
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/catalog/StarOSAgent.java:172]:
```suggestion
        LOG.info("remove worker {} success from StarMgr", workerIpPort);
```

File: fe/fe-core/src/main/java/com/starrocks/catalog/StarOSAgent.java [hunk 6]
```diff
@@ -148,6 +149,30 @@ public void addWorker(long backendId, String workerIpPort) {
         LOG.info("add worker {} success, backendId is {}", workerId, backendId);
     }
 
+    public void removeWorker(String workerIpPort) throws DdlException {
+        long workerId = -1;
+        // get workerId from starMgr
+        try {
+            WorkerInfo workerInfo = client.getWorkerInfo(serviceId, workerIpPort);
+            workerId = workerInfo.getWorkerId();
+        } catch (StarClientException e) {
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/catalog/StarOSAgent.java:158]: Need check ExceptionCode.NOT_EXIST

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 14 / 24 (58.33%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/StarOSAgent.java|10|18|55.56%|
|:large_blue_circle:|com/starrocks/system/SystemInfoService.java|4|6|66.67%|

**Review by wyb — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 14 / 27 (51.85%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/StarOSAgent.java|10|21|47.62%|
|:large_blue_circle:|com/starrocks/system/SystemInfoService.java|4|6|66.67%|

**Comment by abc982627271** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 21 / 27 (77.78%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/system/SystemInfoService.java|4|6|66.67%|
|:large_blue_circle:|com/starrocks/catalog/StarOSAgent.java|17|21|80.95%|

**Review by wyb — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 21 / 27 (77.78%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/system/SystemInfoService.java|4|6|66.67%|
|:large_blue_circle:|com/starrocks/catalog/StarOSAgent.java|17|21|80.95%|

File: fe/fe-core/src/main/java/com/starrocks/catalog/StarOSAgent.java [hunk 7]
```diff
@@ -148,6 +149,37 @@ public void addWorker(long backendId, String workerIpPort) {
...
+            if (e.getCode() != StarClientException.ExceptionCode.NOT_EXIST) {
+                throw new DdlException("Failed to get worker id from starMgr. error: "
+                        + e.getMessage());
+            }
+
+            workerToBackend.remove(workerId);
+            workerToId.remove(workerIpPort);
+            LOG.info("worker {} not exist.", workerIpPort);
+            return;
```
Comment (starrocks-xupeng) [inline fe/fe-core/src/main/java/com/starrocks/catalog/StarOSAgent.java:167]: workerId might not set if exception is thrown.

**Comment by abc982627271** [comment]
run starrocks_fe_unittest

**Comment by abc982627271** [comment]
run starrocks_fe_unittest

**Comment by gengjun-git** [comment]
run starrocks_fe_unittest

**Comment by abc982627271** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 23 / 27 (85.19%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/system/SystemInfoService.java|4|6|66.67%|
|:large_blue_circle:|com/starrocks/catalog/StarOSAgent.java|19|21|90.48%|

**Comment by abc982627271** [comment]
run starrocks_fe_unittest

**Comment by abc982627271** [comment]
run starrocks_fe_unittest

**Comment by abc982627271** [comment]
run starrocks_fe_unittest

**Review by wyb — APPROVED** [review]

**Review by wuleistarrocks — APPROVED** [review]

**Review by sduzh — APPROVED** [review]

**Pull request merged**